lib/configuration.js

```diff
@@ -15,150 +15,429 @@
+ *  retrieved, null otherwise.
+ * @returns {Undefined} - does not return anything
  */
-function gatherConfiguration(initConfiguration) {
-
-  var projectId = null;
-  var reportUncaughtExceptions = true;
-  var serviceContext = null;
-  var key = null;
-  var shouldReportErrorsToAPI = false;
-
-  if (env.NODE_ENV === 'production') {
-
-    shouldReportErrorsToAPI = true;
+Configuration.prototype._assimilateProjectNumber = function(err, projectNum) {
+  if (isNull(err) && isString(projectNum)) {
+    this._projectNumber = projectNum;
   } else {
-
-    console.log([
-      'Warning: the NODE_ENV environmental variable is not set to',
-      '\'production\'. To send errors to the Stackdriver Error Reporting API',
-      'please set the NODE_ENV to \'production\'.'
-    ].join(' '));
+    this._checkLocalProjectNumber();
   }
-
-  if (isPlainObject(initConfiguration)) {
-
-    if (has(initConfiguration, 'projectId')) {
-
-      projectId = initConfiguration.projectId;
-    }
-
-    if (has(initConfiguration, 'reportUncaughtExceptions')) {
-
-      reportUncaughtExceptions = !!initConfiguration.reportUncaughtExceptions;
-    }
-
-    if (has(initConfiguration, 'key')) {
-
-      key = initConfiguration.key;
-    }
-  } else if (!isUndefined(initConfiguration)) {
-
-    throw new Error(
-        'An invalid configuration was given to the Stackdriver Error Library');
+};
+/**
+ * The _checkLocalProjectId function is responsible for determing whether the
+ * _projectId property was set by the metadata service and whether or not the
+ * _projectId property should/can be set with a environmental or runtime
+ * configuration variable. If, upon execution of the _checkLocalProjectId
+ * function, the _projectId property has already been set to a string then it is
+ * assumed that this property has been set with the metadata services response.
+ * The metadata value for the project id always take precedence over any other
+ * locally configured project id value. Given that the metadata service did not
+ * set the project id this function will defer next to the value set in the
+ * environment named `GCLOUD_PROJECT` if it is set and of type string. If this
+ * environmental variable is not set the function will defer to the
+ * _givenConfiguration property if it is of type object and has a string
+ * property named projectId. If none of these conditions are met then the
+ * _projectId property will be left at its default value.
+ * @memberof Configuration
+ * @private
+ * @function _checkLocalProjectId
+ * @returns {Undefined} - does not return anything
+ */
+Configuration.prototype._checkLocalProjectId = function() {
+  if (isString(this._projectId)) {
+    // already has been set by the metadata service
+    // this call is still waiting to go live in the utils library
+    return;
+  } else if (isString(env.GCLOUD_PROJECT) && !isEmpty(env.GCLOUD_PROJECT) &&
+    isNaN(env.GCLOUD_PROJECT)) {
```

Comment:
@ofrobots No this is check to see if it is not a number otherwise, if it is a number, it is considered to be a project id